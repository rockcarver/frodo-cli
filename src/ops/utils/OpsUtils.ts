import { FrodoError } from '@rockcarver/frodo-lib';

import { printError } from '../../utils/Console';

/**
 * Deep clone object
 * @param {any} obj object to deep clone
 * @returns {any} new object cloned from obj
 */
export function cloneDeep(obj: any): any {
  return JSON.parse(JSON.stringify(obj));
}

/**
 * Result callback that handles errors when they occur by printing them and setting the process exit code
 * @param error The error to handle
 */
export function errorHandler(error?: FrodoError) {
  if (!error) return;
  printError(error);
  process.exitCode = 1;
}
