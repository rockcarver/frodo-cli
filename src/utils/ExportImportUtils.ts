import fs from 'fs';
import slugify from 'slugify';
import { ExportImportUtils, state } from '@rockcarver/frodo-lib';
import { printMessage } from './Console';

const { getMetadata } = ExportImportUtils;

/**
 * Get a typed filename. E.g. "my-script.script.json"
 *
 * @param name The name of the file
 * @param type The type of the file, e.g. script, idp, etc.
 * @param suffix The suffix of the file, e.g. json, xml, etc. Defaults to json.
 * @returns The typed filename
 */
export function getTypedFilename(
  name: string,
  type: string,
  suffix = 'json'
): string {
  const slug = slugify(name.replace(/^http(s?):\/\//, ''));
  return `${slug}.${type}.${suffix}`;
}

/**
 * Save JSON object to file
 *
 * @param data data object
 * @param filename file name
 */
export function saveJsonToFile(data: any, filename: string) {
  const exportData = data;
  if (!exportData.meta) {
    exportData.meta = getMetadata();
  }
  fs.writeFile(filename, JSON.stringify(exportData, null, 2), (err) => {
    if (err) {
      return printMessage(`ERROR - can't save ${filename}`, 'error');
    }
    return '';
  });
}

export function saveToFile(type, data, identifier, filename) {
  const exportData = {};
  exportData['meta'] = getMetadata();
  exportData[type] = {};

  if (Array.isArray(data)) {
    data.forEach((element) => {
      exportData[type][element[identifier]] = element;
    });
  } else {
    exportData[type][data[identifier]] = data;
  }
  fs.writeFile(filename, JSON.stringify(exportData, null, 2), (err) => {
    if (err) {
      return printMessage(`ERROR - can't save ${type} to file`, 'error');
    }
    return '';
  });
}

/**
 * Save text data to file
 * @param data text data
 * @param filename file name
 * @return true if successful, false otherwise
 */
export function saveTextToFile(data: string, filename: string): boolean {
  try {
    fs.writeFileSync(filename, data);
    return true;
  } catch (error) {
    printMessage(`ERROR - can't save ${filename}`, 'error');
    return false;
  }
}

/*
 * Output str in title case
 *
 * e.g.: 'ALL UPPERCASE AND all lowercase' = 'All Uppercase And All Lowercase'
 */
export function titleCase(input) {
  const str = input.toString();
  const splitStr = str.toLowerCase().split(' ');
  for (let i = 0; i < splitStr.length; i += 1) {
    splitStr[i] = splitStr[i].charAt(0).toUpperCase() + splitStr[i].slice(1);
  }
  return splitStr.join(' ');
}

export function getRealmString() {
  const realm = state.getRealm();
  return realm
    .split('/')
    .reduce((result, item) => `${result}${titleCase(item)}`, '');
}
