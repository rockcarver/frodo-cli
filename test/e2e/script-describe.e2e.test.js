/**
 * Follow this process to write e2e tests for the CLI project:
 *
 * 1. Test if all the necessary mocks for your tests already exist.
 *    In mock mode, run the command you want to test with the same arguments
 *    and parameters exactly as you want to test it, for example:
 *
 *    $ FRODO_MOCK=1 frodo conn save https://openam-frodo-dev.forgeblocks.com/am volker.scheuber@forgerock.com Sup3rS3cr3t!
 *
 *    If your command completes without errors and with the expected results,
 *    all the required mocks already exist and you are good to write your
 *    test and skip to step #4.
 *
 *    If, however, your command fails and you see errors like the one below,
 *    you know you need to record the mock responses first:
 *
 *    [Polly] [adapter:node-http] Recording for the following request is not found and `recordIfMissing` is `false`.
 *
 * 2. Record mock responses for your exact command.
 *    In mock record mode, run the command you want to test with the same arguments
 *    and parameters exactly as you want to test it, for example:
 *
 *    $ FRODO_MOCK=record frodo conn save https://openam-frodo-dev.forgeblocks.com/am volker.scheuber@forgerock.com Sup3rS3cr3t!
 *
 *    Wait until you see all the Polly instances (mock recording adapters) have
 *    shutdown before you try to run step #1 again.
 *    Messages like these indicate mock recording adapters shutting down:
 *
 *    Polly instance 'conn/4' stopping in 3s...
 *    Polly instance 'conn/4' stopping in 2s...
 *    Polly instance 'conn/save/3' stopping in 3s...
 *    Polly instance 'conn/4' stopping in 1s...
 *    Polly instance 'conn/save/3' stopping in 2s...
 *    Polly instance 'conn/4' stopped.
 *    Polly instance 'conn/save/3' stopping in 1s...
 *    Polly instance 'conn/save/3' stopped.
 *
 * 3. Validate your freshly recorded mock responses are complete and working.
 *    Re-run the exact command you want to test in mock mode (see step #1).
 *
 * 4. Write your test.
 *    Make sure to use the exact command including number of arguments and params.
 *
 * 5. Commit both your test and your new recordings to the repository.
 *    Your tests are likely going to reside outside the frodo-lib project but
 *    the recordings must be committed to the frodo-lib project.
 */

/*
FRODO_MOCK=record FRODO_NO_CACHE=1 FRODO_HOST=https://openam-frodo-dev.forgeblocks.com/am frodo script describe -i 1b52a7e0-4019-40fa-958a-15a49870e901
FRODO_MOCK=record FRODO_NO_CACHE=1 FRODO_HOST=https://openam-frodo-dev.forgeblocks.com/am frodo script describe -un shared
FRODO_MOCK=record FRODO_NO_CACHE=1 FRODO_HOST=https://openam-frodo-dev.forgeblocks.com/am frodo script describe -un shared -f test/e2e/exports/all/all.cloud.json
FRODO_MOCK=record FRODO_NO_CACHE=1 FRODO_HOST=https://openam-frodo-dev.forgeblocks.com/am frodo script describe -un shared -D test/e2e/exports/all-separate/cloud
FRODO_MOCK=record FRODO_NO_CACHE=1 FRODO_HOST=https://openam-frodo-dev.forgeblocks.com/am frodo script describe --json --usage --script-id 1b52a7e0-4019-40fa-958a-15a49870e901 --file test/e2e/exports/all/all.cloud.json
FRODO_MOCK=record FRODO_NO_CACHE=1 FRODO_HOST=https://openam-frodo-dev.forgeblocks.com/am frodo script describe --json --usage --script-name shared --directory test/e2e/exports/all-separate/cloud
*/
import cp from 'child_process';
import { promisify } from 'util';
import { getEnv, removeAnsiEscapeCodes, removeProgressBarOutput } from './utils/TestUtils';
import { connection as c } from './utils/TestConfig';

const exec = promisify(cp.exec);

process.env['FRODO_MOCK'] = '1';
const env = getEnv(c);

const allConfigFile = 'test/e2e/exports/all/all.cloud.json';
const allConfigDirectory = 'test/e2e/exports/all-separate/cloud';

describe('frodo script describe', () => {
    test(`"frodo script describe -i 1b52a7e0-4019-40fa-958a-15a49870e901": should describe the script with id "1b52a7e0-4019-40fa-958a-15a49870e901"`, async () => {
        const CMD = `frodo script describe -i 1b52a7e0-4019-40fa-958a-15a49870e901`;
        const { stdout } = await exec(CMD, env);
        expect(removeAnsiEscapeCodes(stdout)).toMatchSnapshot();
    });

    test(`"frodo script describe -un shared": should describe the script with name "shared" with usage`, async () => {
        const CMD = `frodo script describe -un shared`;
        try {
            await exec(CMD, env);
            fail("Command should've failed")
        } catch (e) {
            expect(removeProgressBarOutput(removeAnsiEscapeCodes(e.stderr))).toMatchSnapshot();
            expect(removeProgressBarOutput(removeAnsiEscapeCodes(e.stdout))).toMatchSnapshot();
        }
    });

    test(`"frodo script describe -un shared -f ${allConfigFile}": should describe the script with name "shared" with usage from file ${allConfigFile}`, async () => {
        const CMD = `frodo script describe -un shared -f ${allConfigFile}`;
        const { stdout } = await exec(CMD, env);
        expect(removeAnsiEscapeCodes(stdout)).toMatchSnapshot();
    });

    test(`"frodo script describe -un shared -D ${allConfigDirectory}": should describe the script with name "shared" with usage from directory ${allConfigDirectory}`, async () => {
        const CMD = `frodo script describe -un shared -D ${allConfigDirectory}`;
        const { stdout } = await exec(CMD, env);
        expect(removeAnsiEscapeCodes(stdout)).toMatchSnapshot();
    });

    test(`"frodo script describe --json --usage --script-id 1b52a7e0-4019-40fa-958a-15a49870e901 --file ${allConfigFile}": should describe the script with id "1b52a7e0-4019-40fa-958a-15a49870e901" with usage from file ${allConfigFile} and json output`, async () => {
        const CMD = `frodo script describe --json --usage --script-id 1b52a7e0-4019-40fa-958a-15a49870e901 --file ${allConfigFile}`;
        const { stdout } = await exec(CMD, env);
        expect(removeAnsiEscapeCodes(stdout)).toMatchSnapshot();
    });

    test(`"frodo script describe --json --usage --script-name shared --directory ${allConfigDirectory}": should describe the script with name "shared" with usage from directory ${allConfigDirectory} and json output`, async () => {
        const CMD = `frodo script describe --json --usage --script-name shared --directory ${allConfigDirectory}`;
        const { stdout } = await exec(CMD, env);
        expect(removeAnsiEscapeCodes(stdout)).toMatchSnapshot();
    });
});
