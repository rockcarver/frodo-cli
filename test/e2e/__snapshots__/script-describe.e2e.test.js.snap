// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`frodo script describe "frodo script describe --json --usage --script-id 1b52a7e0-4019-40fa-958a-15a49870e901 --file test/e2e/exports/all/all.cloud.json": should describe the script with id "1b52a7e0-4019-40fa-958a-15a49870e901" with usage from file test/e2e/exports/all/all.cloud.json and json output 1`] = `
"{
  _id: '1b52a7e0-4019-40fa-958a-15a49870e901',
  name: 'shared',
  description: 'set the same shared state variable',
  script: 'KGZ1bmN0aW9uICgpIHsKICBvdXRjb21lID0gJ3RydWUnOwogIHZhciBsZXZlbCA9IG5vZGVTdGF0ZS5nZXQoJ2xldmVsJykuYXNJbnRlZ2VyKCk7CiAgc2hhcmVkU3RhdGUucHV0KCdzaGFyZWRWYWx1ZScsICdMZXZlbCAnICsgbGV2ZWwgKyAnOiBUaGlzIGlzIGEgbG9uZ2VyIHN0cmluZyB2YWx1ZSBzaGFyZWQgYWNyb3NzIGFsbCBuZXN0ZWQgam91cm5leXMuIEl0IGNvbnRhaW5zIGFuIGluZGljYXRvciBpbiB3aGljaCBsZXZlbCBpdCB3YXMgbGFzdCBzZXQuJyk7Cn0oKSk7',
  default: false,
  language: 'JAVASCRIPT',
  context: 'AUTHENTICATION_TREE_DECISION_NODE',
  createdBy: 'null',
  creationDate: 0,
  lastModifiedBy: 'null',
  lastModifiedDate: 0,
  evaluatorVersion: '1.0',
  locations: [
    'realm.root-alpha.trees.j00.nodes.01d3785f-7fb4-44a7-9458-72c380a9818f.script',
    'realm.root-alpha.trees.j00.nodes.d17ffaa1-2c61-4abd-9bb1-2559160d0a5c.script',
    'realm.root-alpha.trees.j01.nodes.6674b4ac-dd89-4e13-9440-6f81194e3a22.script',
    'realm.root-alpha.trees.j01.nodes.89ce5d57-82fa-4d58-8d15-0329f7dbd7e7.script',
    'realm.root-alpha.trees.j02.nodes.cbb3d506-b267-4b99-9edd-363e90aac997.script',
    'realm.root-alpha.trees.j02.nodes.e0983ead-4918-48f6-858d-9aff0f03759c.script',
    'realm.root-alpha.trees.j03.nodes.6f9de973-9ed4-41f5-b43d-4036041e2b96.script',
    'realm.root-alpha.trees.j03.nodes.fae7424e-13c9-45bd-b3a2-045773671a3f.script',
    'realm.root-alpha.trees.j04.nodes.9603ef52-30f0-4ddc-b3c0-28dac83c7bdb.script',
    'realm.root-alpha.trees.j04.nodes.f5c317ce-fabd-4a10-9907-c71cea037844.script',
    'realm.root-alpha.trees.j05.nodes.11f1c31c-50a9-4717-8213-420f6932481f.script',
    'realm.root-alpha.trees.j05.nodes.a0782616-84b7-4bf5-87ed-a01fb3018563.script',
    'realm.root-alpha.trees.j06.nodes.1d59caff-243c-45bd-b7d0-6dcc563989c5.script',
    'realm.root-alpha.trees.j06.nodes.da878771-421c-463f-aad7-4d5f2ad5e59a.script',
    'realm.root-alpha.trees.j07.nodes.ac6ee166-73c0-4f73-b8db-4fe8ff6a25c0.script',
    'realm.root-alpha.trees.j07.nodes.d9a06d3a-7e3f-4244-9a32-63ffa0d26e00.script',
    'realm.root-alpha.trees.j08.nodes.042b600b-71cb-45a8-93ae-a6f57b16a6e5.script',
    'realm.root-alpha.trees.j08.nodes.948e21f4-c512-450a-9d42-e0d629217834.script',
    'realm.root-alpha.trees.j09.nodes.8c5e9cb5-471b-4dd6-b150-ecaaeda98195.script',
    'realm.root-alpha.trees.j09.nodes.f57cf53c-b4c6-48f7-84e8-91f535a2e8f8.script',
    'realm.root-alpha.trees.j10.nodes.40afb384-e9b6-4dcb-acde-04de109474c8.script',
    'realm.root-alpha.trees.j10.nodes.97ef9d96-99e7-4d2d-b6c6-4177b5397ead.script'
  ]
}
"
`;

exports[`frodo script describe "frodo script describe --json --usage --script-name shared --directory test/e2e/exports/all-separate/cloud": should describe the script with name "shared" with usage from directory test/e2e/exports/all-separate/cloud and json output 1`] = `
"{
  _id: '1b52a7e0-4019-40fa-958a-15a49870e901',
  name: 'shared',
  description: 'set the same shared state variable',
  script: 'KGZ1bmN0aW9uICgpIHsKICBvdXRjb21lID0gJ3RydWUnOwogIHZhciBsZXZlbCA9IG5vZGVTdGF0ZS5nZXQoJ2xldmVsJykuYXNJbnRlZ2VyKCk7CiAgc2hhcmVkU3RhdGUucHV0KCdzaGFyZWRWYWx1ZScsICdMZXZlbCAnICsgbGV2ZWwgKyAnOiBUaGlzIGlzIGEgbG9uZ2VyIHN0cmluZyB2YWx1ZSBzaGFyZWQgYWNyb3NzIGFsbCBuZXN0ZWQgam91cm5leXMuIEl0IGNvbnRhaW5zIGFuIGluZGljYXRvciBpbiB3aGljaCBsZXZlbCBpdCB3YXMgbGFzdCBzZXQuJyk7Cn0oKSk7',
  default: false,
  language: 'JAVASCRIPT',
  context: 'AUTHENTICATION_TREE_DECISION_NODE',
  createdBy: 'null',
  creationDate: 0,
  lastModifiedBy: 'null',
  lastModifiedDate: 0,
  evaluatorVersion: '1.0',
  locations: [
    'realm.root-alpha.trees.j00.nodes.01d3785f-7fb4-44a7-9458-72c380a9818f.script',
    'realm.root-alpha.trees.j00.nodes.d17ffaa1-2c61-4abd-9bb1-2559160d0a5c.script',
    'realm.root-alpha.trees.j01.nodes.6674b4ac-dd89-4e13-9440-6f81194e3a22.script',
    'realm.root-alpha.trees.j01.nodes.89ce5d57-82fa-4d58-8d15-0329f7dbd7e7.script',
    'realm.root-alpha.trees.j02.nodes.cbb3d506-b267-4b99-9edd-363e90aac997.script',
    'realm.root-alpha.trees.j02.nodes.e0983ead-4918-48f6-858d-9aff0f03759c.script',
    'realm.root-alpha.trees.j03.nodes.6f9de973-9ed4-41f5-b43d-4036041e2b96.script',
    'realm.root-alpha.trees.j03.nodes.fae7424e-13c9-45bd-b3a2-045773671a3f.script',
    'realm.root-alpha.trees.j04.nodes.9603ef52-30f0-4ddc-b3c0-28dac83c7bdb.script',
    'realm.root-alpha.trees.j04.nodes.f5c317ce-fabd-4a10-9907-c71cea037844.script',
    'realm.root-alpha.trees.j05.nodes.11f1c31c-50a9-4717-8213-420f6932481f.script',
    'realm.root-alpha.trees.j05.nodes.a0782616-84b7-4bf5-87ed-a01fb3018563.script',
    'realm.root-alpha.trees.j06.nodes.1d59caff-243c-45bd-b7d0-6dcc563989c5.script',
    'realm.root-alpha.trees.j06.nodes.da878771-421c-463f-aad7-4d5f2ad5e59a.script',
    'realm.root-alpha.trees.j07.nodes.ac6ee166-73c0-4f73-b8db-4fe8ff6a25c0.script',
    'realm.root-alpha.trees.j07.nodes.d9a06d3a-7e3f-4244-9a32-63ffa0d26e00.script',
    'realm.root-alpha.trees.j08.nodes.042b600b-71cb-45a8-93ae-a6f57b16a6e5.script',
    'realm.root-alpha.trees.j08.nodes.948e21f4-c512-450a-9d42-e0d629217834.script',
    'realm.root-alpha.trees.j09.nodes.8c5e9cb5-471b-4dd6-b150-ecaaeda98195.script',
    'realm.root-alpha.trees.j09.nodes.f57cf53c-b4c6-48f7-84e8-91f535a2e8f8.script',
    'realm.root-alpha.trees.j10.nodes.40afb384-e9b6-4dcb-acde-04de109474c8.script',
    'realm.root-alpha.trees.j10.nodes.97ef9d96-99e7-4d2d-b6c6-4177b5397ead.script'
  ]
}
"
`;

exports[`frodo script describe "frodo script describe -i 1b52a7e0-4019-40fa-958a-15a49870e901": should describe the script with id "1b52a7e0-4019-40fa-958a-15a49870e901" 1`] = `
"Id               │1b52a7e0-4019-40fa-958a-15a49870e901                                             
Name             │shared                                                                           
Language         │JavaScript                                                                       
Context          │Authentication Tree Decision Node                                                
Description      │set the same shared state variable                                               
Default          │false                                                                            
Evaluator Version│1.0                                                                              
Script (Base 64) │KGZ1bmN0aW9uICgpIHsKICBvdXRjb21lID0gJ3RydWUnOwogIHZhciBsZXZlbCA9IG5vZGVTdGF0ZS5nZ
                 │XQoJ2xldmVsJykuYXNJbnRlZ2VyKCk7CiAgc2hhcmVkU3RhdGUucHV0KCdzaGFyZWRWYWx1ZScsICdMZX
                 │ZlbCAnICsgbGV2ZWwgKyAnOiBUaGlzIGlzIGEgbG9uZ2VyIHN0cmluZyB2YWx1ZSBzaGFyZWQgYWNyb3N
                 │zIGFsbCBuZXN0ZWQgam91cm5leXMuIEl0IGNvbnRhaW5zIGFuIGluZGljYXRvciBpbiB3aGljaCBsZXZl
                 │bCBpdCB3YXMgbGFzdCBzZXQuJyk7Cn0oKSk7                                             
"
`;

exports[`frodo script describe "frodo script describe -un shared -D test/e2e/exports/all-separate/cloud": should describe the script with name "shared" with usage from directory test/e2e/exports/all-separate/cloud 1`] = `
"Id                        │1b52a7e0-4019-40fa-958a-15a49870e901                                             
Name                      │shared                                                                           
Language                  │JavaScript                                                                       
Context                   │Authentication Tree Decision Node                                                
Description               │set the same shared state variable                                               
Default                   │false                                                                            
Evaluator Version         │1.0                                                                              
Script (Base 64)          │KGZ1bmN0aW9uICgpIHsKICBvdXRjb21lID0gJ3RydWUnOwogIHZhciBsZXZlbCA9IG5vZGVTdGF0ZS5nZ
                          │XQoJ2xldmVsJykuYXNJbnRlZ2VyKCk7CiAgc2hhcmVkU3RhdGUucHV0KCdzaGFyZWRWYWx1ZScsICdMZX
                          │ZlbCAnICsgbGV2ZWwgKyAnOiBUaGlzIGlzIGEgbG9uZ2VyIHN0cmluZyB2YWx1ZSBzaGFyZWQgYWNyb3N
                          │zIGFsbCBuZXN0ZWQgam91cm5leXMuIEl0IGNvbnRhaW5zIGFuIGluZGljYXRvciBpbiB3aGljaCBsZXZl
                          │bCBpdCB3YXMgbGFzdCBzZXQuJyk7Cn0oKSk7                                             
Usage Locations (22 total)│realm.root-alpha.trees.j00.nodes.01d3785f-7fb4-44a7-9458-72c380a9818f.script     
                          │realm.root-alpha.trees.j00.nodes.d17ffaa1-2c61-4abd-9bb1-2559160d0a5c.script     
                          │realm.root-alpha.trees.j01.nodes.6674b4ac-dd89-4e13-9440-6f81194e3a22.script     
                          │realm.root-alpha.trees.j01.nodes.89ce5d57-82fa-4d58-8d15-0329f7dbd7e7.script     
                          │realm.root-alpha.trees.j02.nodes.cbb3d506-b267-4b99-9edd-363e90aac997.script     
                          │realm.root-alpha.trees.j02.nodes.e0983ead-4918-48f6-858d-9aff0f03759c.script     
                          │realm.root-alpha.trees.j03.nodes.6f9de973-9ed4-41f5-b43d-4036041e2b96.script     
                          │realm.root-alpha.trees.j03.nodes.fae7424e-13c9-45bd-b3a2-045773671a3f.script     
                          │realm.root-alpha.trees.j04.nodes.9603ef52-30f0-4ddc-b3c0-28dac83c7bdb.script     
                          │realm.root-alpha.trees.j04.nodes.f5c317ce-fabd-4a10-9907-c71cea037844.script     
                          │realm.root-alpha.trees.j05.nodes.11f1c31c-50a9-4717-8213-420f6932481f.script     
                          │realm.root-alpha.trees.j05.nodes.a0782616-84b7-4bf5-87ed-a01fb3018563.script     
                          │realm.root-alpha.trees.j06.nodes.1d59caff-243c-45bd-b7d0-6dcc563989c5.script     
                          │realm.root-alpha.trees.j06.nodes.da878771-421c-463f-aad7-4d5f2ad5e59a.script     
                          │realm.root-alpha.trees.j07.nodes.ac6ee166-73c0-4f73-b8db-4fe8ff6a25c0.script     
                          │realm.root-alpha.trees.j07.nodes.d9a06d3a-7e3f-4244-9a32-63ffa0d26e00.script     
                          │realm.root-alpha.trees.j08.nodes.042b600b-71cb-45a8-93ae-a6f57b16a6e5.script     
                          │realm.root-alpha.trees.j08.nodes.948e21f4-c512-450a-9d42-e0d629217834.script     
                          │realm.root-alpha.trees.j09.nodes.8c5e9cb5-471b-4dd6-b150-ecaaeda98195.script     
                          │realm.root-alpha.trees.j09.nodes.f57cf53c-b4c6-48f7-84e8-91f535a2e8f8.script     
                          │realm.root-alpha.trees.j10.nodes.40afb384-e9b6-4dcb-acde-04de109474c8.script     
                          │realm.root-alpha.trees.j10.nodes.97ef9d96-99e7-4d2d-b6c6-4177b5397ead.script     
"
`;

exports[`frodo script describe "frodo script describe -un shared -f test/e2e/exports/all/all.cloud.json": should describe the script with name "shared" with usage from file test/e2e/exports/all/all.cloud.json 1`] = `
"Id                        │1b52a7e0-4019-40fa-958a-15a49870e901                                             
Name                      │shared                                                                           
Language                  │JavaScript                                                                       
Context                   │Authentication Tree Decision Node                                                
Description               │set the same shared state variable                                               
Default                   │false                                                                            
Evaluator Version         │1.0                                                                              
Script (Base 64)          │KGZ1bmN0aW9uICgpIHsKICBvdXRjb21lID0gJ3RydWUnOwogIHZhciBsZXZlbCA9IG5vZGVTdGF0ZS5nZ
                          │XQoJ2xldmVsJykuYXNJbnRlZ2VyKCk7CiAgc2hhcmVkU3RhdGUucHV0KCdzaGFyZWRWYWx1ZScsICdMZX
                          │ZlbCAnICsgbGV2ZWwgKyAnOiBUaGlzIGlzIGEgbG9uZ2VyIHN0cmluZyB2YWx1ZSBzaGFyZWQgYWNyb3N
                          │zIGFsbCBuZXN0ZWQgam91cm5leXMuIEl0IGNvbnRhaW5zIGFuIGluZGljYXRvciBpbiB3aGljaCBsZXZl
                          │bCBpdCB3YXMgbGFzdCBzZXQuJyk7Cn0oKSk7                                             
Usage Locations (22 total)│realm.root-alpha.trees.j00.nodes.01d3785f-7fb4-44a7-9458-72c380a9818f.script     
                          │realm.root-alpha.trees.j00.nodes.d17ffaa1-2c61-4abd-9bb1-2559160d0a5c.script     
                          │realm.root-alpha.trees.j01.nodes.6674b4ac-dd89-4e13-9440-6f81194e3a22.script     
                          │realm.root-alpha.trees.j01.nodes.89ce5d57-82fa-4d58-8d15-0329f7dbd7e7.script     
                          │realm.root-alpha.trees.j02.nodes.cbb3d506-b267-4b99-9edd-363e90aac997.script     
                          │realm.root-alpha.trees.j02.nodes.e0983ead-4918-48f6-858d-9aff0f03759c.script     
                          │realm.root-alpha.trees.j03.nodes.6f9de973-9ed4-41f5-b43d-4036041e2b96.script     
                          │realm.root-alpha.trees.j03.nodes.fae7424e-13c9-45bd-b3a2-045773671a3f.script     
                          │realm.root-alpha.trees.j04.nodes.9603ef52-30f0-4ddc-b3c0-28dac83c7bdb.script     
                          │realm.root-alpha.trees.j04.nodes.f5c317ce-fabd-4a10-9907-c71cea037844.script     
                          │realm.root-alpha.trees.j05.nodes.11f1c31c-50a9-4717-8213-420f6932481f.script     
                          │realm.root-alpha.trees.j05.nodes.a0782616-84b7-4bf5-87ed-a01fb3018563.script     
                          │realm.root-alpha.trees.j06.nodes.1d59caff-243c-45bd-b7d0-6dcc563989c5.script     
                          │realm.root-alpha.trees.j06.nodes.da878771-421c-463f-aad7-4d5f2ad5e59a.script     
                          │realm.root-alpha.trees.j07.nodes.ac6ee166-73c0-4f73-b8db-4fe8ff6a25c0.script     
                          │realm.root-alpha.trees.j07.nodes.d9a06d3a-7e3f-4244-9a32-63ffa0d26e00.script     
                          │realm.root-alpha.trees.j08.nodes.042b600b-71cb-45a8-93ae-a6f57b16a6e5.script     
                          │realm.root-alpha.trees.j08.nodes.948e21f4-c512-450a-9d42-e0d629217834.script     
                          │realm.root-alpha.trees.j09.nodes.8c5e9cb5-471b-4dd6-b150-ecaaeda98195.script     
                          │realm.root-alpha.trees.j09.nodes.f57cf53c-b4c6-48f7-84e8-91f535a2e8f8.script     
                          │realm.root-alpha.trees.j10.nodes.40afb384-e9b6-4dcb-acde-04de109474c8.script     
                          │realm.root-alpha.trees.j10.nodes.97ef9d96-99e7-4d2d-b6c6-4177b5397ead.script     
"
`;

exports[`frodo script describe "frodo script describe -un shared": should describe the script with name "shared" with usage 1`] = `
"Connected to https://openam-frodo-dev.forgeblocks.com/am [alpha] as service account Frodo-SA-1720799681233 [b672336b-41ef-428d-ae4a-e0c082875377]
Error exporting idm config entity fidc/federation-EntraID
  Error reading config entity fidc/federation-EntraID
  HTTP client error
    Code: ERR_BAD_REQUEST
    Status: 403
    Reason: Forbidden
    Message: Access denied
"
`;

exports[`frodo script describe "frodo script describe -un shared": should describe the script with name "shared" with usage 2`] = `
"Id                        │1b52a7e0-4019-40fa-958a-15a49870e901                                             
Name                      │shared                                                                           
Language                  │JavaScript                                                                       
Context                   │Authentication Tree Decision Node                                                
Description               │set the same shared state variable                                               
Default                   │false                                                                            
Evaluator Version         │1.0                                                                              
Script (Base 64)          │KGZ1bmN0aW9uICgpIHsKICBvdXRjb21lID0gJ3RydWUnOwogIHZhciBsZXZlbCA9IG5vZGVTdGF0ZS5nZ
                          │XQoJ2xldmVsJykuYXNJbnRlZ2VyKCk7CiAgc2hhcmVkU3RhdGUucHV0KCdzaGFyZWRWYWx1ZScsICdMZX
                          │ZlbCAnICsgbGV2ZWwgKyAnOiBUaGlzIGlzIGEgbG9uZ2VyIHN0cmluZyB2YWx1ZSBzaGFyZWQgYWNyb3N
                          │zIGFsbCBuZXN0ZWQgam91cm5leXMuIEl0IGNvbnRhaW5zIGFuIGluZGljYXRvciBpbiB3aGljaCBsZXZl
                          │bCBpdCB3YXMgbGFzdCBzZXQuJyk7Cn0oKSk7Cg==                                         
Usage Locations (22 total)│realm.root-alpha.trees.j00.nodes.01d3785f-7fb4-44a7-9458-72c380a9818f.script     
                          │realm.root-alpha.trees.j00.nodes.d17ffaa1-2c61-4abd-9bb1-2559160d0a5c.script     
                          │realm.root-alpha.trees.j01.nodes.6674b4ac-dd89-4e13-9440-6f81194e3a22.script     
                          │realm.root-alpha.trees.j01.nodes.89ce5d57-82fa-4d58-8d15-0329f7dbd7e7.script     
                          │realm.root-alpha.trees.j02.nodes.cbb3d506-b267-4b99-9edd-363e90aac997.script     
                          │realm.root-alpha.trees.j02.nodes.e0983ead-4918-48f6-858d-9aff0f03759c.script     
                          │realm.root-alpha.trees.j03.nodes.6f9de973-9ed4-41f5-b43d-4036041e2b96.script     
                          │realm.root-alpha.trees.j03.nodes.fae7424e-13c9-45bd-b3a2-045773671a3f.script     
                          │realm.root-alpha.trees.j04.nodes.9603ef52-30f0-4ddc-b3c0-28dac83c7bdb.script     
                          │realm.root-alpha.trees.j04.nodes.f5c317ce-fabd-4a10-9907-c71cea037844.script     
                          │realm.root-alpha.trees.j05.nodes.11f1c31c-50a9-4717-8213-420f6932481f.script     
                          │realm.root-alpha.trees.j05.nodes.a0782616-84b7-4bf5-87ed-a01fb3018563.script     
                          │realm.root-alpha.trees.j06.nodes.1d59caff-243c-45bd-b7d0-6dcc563989c5.script     
                          │realm.root-alpha.trees.j06.nodes.da878771-421c-463f-aad7-4d5f2ad5e59a.script     
                          │realm.root-alpha.trees.j07.nodes.ac6ee166-73c0-4f73-b8db-4fe8ff6a25c0.script     
                          │realm.root-alpha.trees.j07.nodes.d9a06d3a-7e3f-4244-9a32-63ffa0d26e00.script     
                          │realm.root-alpha.trees.j08.nodes.042b600b-71cb-45a8-93ae-a6f57b16a6e5.script     
                          │realm.root-alpha.trees.j08.nodes.948e21f4-c512-450a-9d42-e0d629217834.script     
                          │realm.root-alpha.trees.j09.nodes.8c5e9cb5-471b-4dd6-b150-ecaaeda98195.script     
                          │realm.root-alpha.trees.j09.nodes.f57cf53c-b4c6-48f7-84e8-91f535a2e8f8.script     
                          │realm.root-alpha.trees.j10.nodes.40afb384-e9b6-4dcb-acde-04de109474c8.script     
                          │realm.root-alpha.trees.j10.nodes.97ef9d96-99e7-4d2d-b6c6-4177b5397ead.script     
"
`;
